test/internal_rules.jl

```diff
@@ -136,327 +157,101 @@ end
-
-        fdJ = op==:\ ? FiniteDifferences.jacobian(fdm, b_one, copy(b))[1] : nothing
-        fwdJ = fwdJdxdb(driver, A, b)
-        revJ = revJdxdb(driver, A, b)
-        batchedrevJ = batchedrevJdxdb(driver, A, b)
-        batchedfwdJ = batchedfwdJdxdb(driver, A, b)
-        J = Jdxdb(driver, A, b)
-
-        if op == :\
-            @test isapprox(fwdJ, fdJ)
-        end
-
-        @test isapprox(fwdJ, revJ)
-        @test isapprox(fwdJ, batchedrevJ)
-        @test isapprox(fwdJ, batchedfwdJ)
-
-        fwdJ = const_fwdJdxdb(driver_NC, cholesky(A), b)
-        revJ = const_revJdxdb(driver_NC, cholesky(A), b)
-        if op == :\
-            @test isapprox(fwdJ, fdJ)
-        end
-        @test isapprox(fwdJ, revJ)
-
-        function h(A, b)
+    @testset "Linear solve for `Cholesky`" begin
+        @testset for Te in (Float64,)
+            A = exp(Symmetric(rand(Te, 4, 4)))
             C = cholesky(A)
-            b2 = copy(b)
-            ldiv!(C, b2)
-            @inbounds b2[1]
+            B = rand(Te, 4, 4)
+            b = rand(Te, 4)
+            @testset for TC in (Const, Duplicated), TB in (Const, Duplicated),
+                Tret in (Const, Duplicated)
+                @testset "$(size(_B))" for _B in (B, b)
+                    are_activities_compatible(Tret, TC, TB) || continue
+                    test_forward(\, Tret, (C, TC), (_B, TB))
+                    test_reverse(\, Tret, (C, TC), (_B, TB))
+                end
+            end
         end
-
-        A = [1.3 0.5; 0.5 1.5]
-        b = [1., 2.]
-        V = [1.0 0.0; 0.0 0.0]
-        dA = zero(A)
-        Enzyme.autodiff(Reverse, h, Active, Duplicated(A, dA), Const(b))
-
-        dA_sym = - (transpose(A) \ [1.0, 0.0]) * transpose(A \ b)
-        @test isapprox(dA, dA_sym)
     end
-end
 
-@testset "Linear solve for triangular matrices" begin
-    @testset for T in (UpperTriangular, LowerTriangular, UnitUpperTriangular, UnitLowerTriangular),
-        TE in (Float64, ComplexF64), sizeB in ((3,), (3, 3))
-        n = sizeB[1]
-        M = rand(TE, n, n)
-        B = rand(TE, sizeB...)
-        Y = zeros(TE, sizeB...)
-        A = T(M)
-        @testset "test through constructor" begin
-            _A = T(A)
-            function f!(Y, A, B, ::T) where T
-                ldiv!(Y, T(A), B)
-                return nothing
+    @testset "Linear solve for triangular matrices" begin
+        @testset for T in (
+                UpperTriangular, LowerTriangular, UnitUpperTriangular, UnitLowerTriangular
+            ),
+            TE in (Float64, ComplexF64),
+            sizeB in ((3,), (3, 3))
+
+            n = sizeB[1]
+            M = rand(TE, n, n)
+            B = rand(TE, sizeB...)
+            Y = zeros(TE, sizeB...)
+            A = T(M)
+            @testset "test through constructor" begin
+                _A = T(A)
+                function f!(Y, A, B, ::T) where {T}
+                    ldiv!(Y, T(A), B)
+                    return nothing
+                end
```

Comment:
EnzymeTestUtils can now handle this, and it's slightly better because it also checks that the rule and primal return the same quantity.

```suggestion
                f!(Y, A, B, ::T) where {T} = ldiv!(Y, T(A), B)
```

Reply:
Those are good points, but they are unrelated to this PR. The only reason they end up as changed lines is because I ran the formatter on these files. I'm going to open another PR for this https://github.com/EnzymeAD/Enzyme.jl/pull/1407